accepted/2023/dotnet-nuget-setapikey.md

```diff
@@ -0,0 +1,234 @@
+In any case, there are servers which require API keys to push packages.
+
+Customers who need to use API keys have two options.
+Either persist the secret in `nuget.config`, so that the `push` command line does not need to specify the secret, or to pass the API key on the `push` command line.
+Passing the API on the command line risks leaking secrets, for example a CI system might not mask the secret in logs, or the secret being persisted in a shell's history file.
+Therefore, risk can be reduced by providing customers commands to add and remove the API to the `NuGet.config` file, especially if it can be done in a way where the secret is not entered into, or output from, the console.
+
+In `nuget.config`, the API keys are in a different section to source credentials.
+There is a command `nuget.exe setapikey` to set the API key in a config file, however, `nuget.exe` does not ship in Visual Studio or the .NET SDK.
+Also, `nuget.exe` requires mono on Linux and Mac, since `nuget.exe` is built for the .NET Framework.
+Therefore, a command in the .NET SDK will significantly reduce the barrier for customers to set values in the `nuget.config` file.
+
+- `dotnet nuget apikey unset`
+
+Customers may want to remove secrets from their `NuGet.config` files, so an `unset` command should be added.
+
+- `nuget.config` support for unencrypted `<apikeys>`
+
+Currently `nuget.config` `<apikeys>` section only supports encrypted values, via .NET's [`System.Security.Cryptography.ProtectedData` APIs](https://learn.microsoft.com/dotnet/api/system.security.cryptography.protecteddata).
+As the important message says on the docs page, on Windows it uses Windows' Data Protection API (DPAPI), which isn't supported on Linux or Mac, and [the .NET libraries team have no plans to support it since it will have different security promises than on Windows](https://github.com/dotnet/runtime/issues/22886).
+There is [an issue to consider another way of having less secure "encrypted" passwords](https://github.com/NuGet/Home/issues/1851) on Linux and Mac (like what Mono does for the `ProtectedData` API), but it hasn't been implemented yet.
+Therefore, in order for this feature not to be blocked, `<apikeys>` needs to support unencrypted (plain text) secrets (just as source credentials do).
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+
+- `dotnet nuget apikey set`
+
+```text
+Description:
+  Set the API key in a NuGet.config file.
+
+Usage:
+  dotnet nuget apikey set <VALUE> [options]
+
+Arguments:
+  <VALUE> The value of the API key. If neither --from-file or --environment-variable is specified, the value is treated as a string literal.
+
+Options:
```

Comment:
nuget.exe setapikey command has `--ForceEnglishOutput` option added to address https://github.com/NuGet/Home/issues/2352 customer ask. I don't see that option in this command.

Reply:
The `dotnet` CLI already has a method to change the output language, [the `DOTNET_CLI_UI_LANGUAGE` environment varliable](https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-environment-variables#dotnet_cli_ui_language), which Fernando recently added support for in our tool commands: https://github.com/NuGet/NuGet.Client/pull/4816

So, I don't want to have an option for one command that does not exist in all the other commands, especially when there's already a supported way to achieve the same thing.